airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/report_streams.py

```diff
@@ -0,0 +1,302 @@
+from pydantic import BaseModel
+from source_amazon_ads.common import SourceContext
+from source_amazon_ads.schemas import JSModel, MetricsReport
+from source_amazon_ads.spec import Spec
+from source_amazon_ads.streams.common import BasicAmazonAdsStream
+
+logger = AirbyteLogger()
+
+
+class RecordType(str, Enum):
+    CAMPAIGNS = "campaigns"
+    ADGROUPS = "adGroups"
+    PRODUCTADS = "productAds"
+    TARGETS = "targets"
+    ASINS = "asins"
+
+
+class Status(str, Enum):
+    IN_PROGRESS = "IN_PROGRESS"
+    SUCCESS = "SUCCESS"
+    FAILURE = "FAILURE"
+
+
+class ReportInitResponse(BaseModel):
+    reportId: str
+    status: str
+
+
+class ReportStatus(BaseModel):
+    status: str
+    location: Optional[str]
+
+
+@dataclass
+class ReportInfo:
+    report_id: str
+    profile_id: int
+    record_type: str
+
+
+class TooManyRequests(Exception):
+    """
+    Custom exception occured when response with 429 status code received
+    """
+
+
+class ReportStream(BasicAmazonAdsStream, ABC):
+    """
+    Common base class for report streams
+    """
+
+    primary_key = None
+    CHECK_INTERVAL_SECONDS = 30
+    REPORT_WAIT_TIMEOUT = timedelta(hours=1)
+    REPORT_DATE_FORMAT = "%Y%m%d"
+    cursor_field = "reportDate"
+
+    def __init__(self, config: Spec, context: SourceContext, authenticator: Oauth2Authenticator):
+        self._authenticator = authenticator
+        self._session = requests.Session()
+        self._generate_model()
+        super().__init__(config, context)
+
+    @property
+    def model(self) -> JSModel:
+        return self._model
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+
+        if not stream_slice:
+            return
+        report_date = stream_slice[self.cursor_field]
+        report_infos = self._init_reports(report_date)
+        logger.info(f"Waiting for {len(report_infos)} report(s) to be generated")
+        start_time_point = datetime.now()
+        while report_infos and datetime.now() <= start_time_point + self.REPORT_WAIT_TIMEOUT:
+            completed_reports = []
+            logger.info(f"Checking report status, {len(report_infos)} report(s) remained")
+            for report_info in report_infos:
+                report_status, download_url = self._check_status(report_info)
+                if report_status == Status.FAILURE:
+                    logger.warn(f"Report for {report_info.profile_id} with {report_info.record_type} type generation failed")
+                    completed_reports.append(report_info)
+                elif report_status == Status.SUCCESS:
+                    metric_objects = self._download_report(report_info, download_url)
+                    for metric_object in metric_objects:
+                        yield self._model(
```

Comment:
cursor_field makes it an incremental stream, but we are reading reports in a not particular way (younger reports might be ready earlier and/or older reports might fail). this will update the state incorrectly

Reply:
All the reports that we waiting for this method belong to date specified in stream_slice, there is no younger/older reports for single read_records call.